### PR TITLE
Reduce calls to git cat-file -s

### PR DESCRIPTION
**Pull request by zeripath**
There are multiple places where there are repeated calls to `git cat-file -s` due to the blobs not being created with their size.

Through judicious use of `git ls-tree -l` and slight adjustments to the indexer code we can avoid a lot of these calls.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: modules/git/parse_nogogit_test.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+		},
+		{
+			Input: `100644 blob ea0d83c9081af9500ac9f804101b3fd0a5c293af	README.md
+100644 blob 037f27dc9d353ae4fd50f0474b2194c593914e35	README_ZH.md
+040000 tree 84b90550547016f73c5dd3f50dea662389e67b6d	assets
+100644 blob 9846a94f7e8350a916632929d0fda38c90dd2ca8	atters.md
+`,
+			Expected: []*TreeEntry{
+				{
+					ID:        MustIDFromString("ea0d83c9081af9500ac9f804101b3fd0a5c293af"),
```
Comment (6543) [inline modules/git/parse_nogogit_test.go]: I'm not sure if we shold test against main repo ...

Comment (zeripath) [reply modules/git/parse_nogogit_test.go]: it's not testing against the main repo - these are synthetic tests that just so happen to have been created from the main repo.

Comment (6543) [reply modules/git/parse_nogogit_test.go]: Wont it change if Readme changes too?

Comment (zeripath) [reply modules/git/parse_nogogit_test.go]: it doesn't look at files - it's just testing the string.

**Comment by zeripath** [comment]
Coupled with #14685 this will reduce the number of git calls in language settings from 2 per file + 2 (git ls-tree & get commit)  -> 2 (git ls-tree and a single semi long running batch.)

**Review by 6543 — APPROVED** [review]

File: modules/git/parse_nogogit_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+					ID:        MustIDFromString("9846a94f7e8350a916632929d0fda38c90dd2ca8"),
+					name:      "SECURITY.md",
+					entryMode: EntryModeBlob,
+					size:      429,
+					sized:     true,
+				},
+				{
+					ID:        MustIDFromString("84b90550547016f73c5dd3f50dea662389e67b6d"),
+					name:      "assets",
+					entryMode: EntryModeTree,
```
Comment (zeripath) [inline modules/git/parse_nogogit_test.go:52]:
```suggestion
					entryMode: EntryModeTree,
					sized: true,
```

File: modules/git/parse_gogit_test.go [hunk 3]
```diff
@@ -47,6 +49,8 @@ func TestParseTreeEntries(t *testing.T) {
 						Name: "example/\n.txt",
 						Mode: filemode.Symlink,
 					},
+					size:  234131,
+					sized: true,
 				},
 				{
 					ID: MustIDFromString("1d01fb729fb0db5881daaa6030f9f2d3cd3d5ae8"),
```
Comment (zeripath) [inline modules/git/parse_gogit_test.go]:
```suggestion
					ID: MustIDFromString("1d01fb729fb0db5881daaa6030f9f2d3cd3d5ae8"),
					sized: true,
```

**Comment by 6543** [comment]
@zeripath need `make fmt`

File: modules/git/parse_nogogit_test.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+					sized: true,
```
Comment (zeripath) [inline modules/git/parse_nogogit_test.go]:
```suggestion
					sized: true,
```

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Comment by zeripath** [comment]
Ok I'll take a look at the bleve and elastic indexers next - those could do with a single batch cat-file too

**Pull request merged**